### Carga de Librerias

In [7]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler

### Carga de DF

In [8]:
bmw = pd.read_csv("/Users/yeraysouto/Desktop/M3/Entregable_1/Data_BMW/bmw_pricing.csv")

## Exploracion de datos

In [9]:
bmw

marca             modelo        km  potencia fecha_registro  \
0      NaN                118  140411.0     100.0     2012-02-01   
1      BMW                 M4   13929.0     317.0            NaN   
2      BMW                320  183297.0     120.0     2012-04-01   
3      BMW                420  128035.0     135.0            NaN   
4      BMW                425   97097.0     160.0            NaN   
...    ...                ...       ...       ...            ...   
4838   BMW    218 Gran Tourer   39743.0     110.0            NaN   
4839   BMW  218 Active Tourer   49832.0     100.0     2015-06-01   
4840   BMW    218 Gran Tourer   19633.0     110.0     2015-10-01   
4841   BMW  218 Active Tourer   27920.0     110.0     2016-04-01   
4842   BMW                525  195840.0     160.0            NaN   

     tipo_gasolina   color   tipo_coche volante_regulable aire_acondicionado  \
0           diesel   black          NaN              True               True   
1           petrol    grey  convertible              True               True   
2           diesel   white          NaN             False              False   
3           diesel     red  convertible              True               True   
4           diesel  silver          NaN              True               True   
...            ...     ...          ...               ...                ...   
4838        diesel   black          NaN             False               True   
4839        diesel    grey          NaN             False               True   
4840        diesel    grey          van             False               True   
4841        diesel   brown          van              True               True   
4842        diesel    grey          NaN              True                NaN   

     camara_trasera asientos_traseros_plegables elevalunas_electrico  \
0             False                         NaN                 True   
1             False                         NaN                False   
2             False                         NaN                 True   
3             False                         NaN                 True   
4             False                       False                False   
...             ...                         ...                  ...   
4838          False                         NaN                False   
4839          False                         NaN                False   
4840          False                         NaN                False   
4841          False                       False                False   
4842          False                         NaN                 True   

     bluetooth   gps alerta_lim_velocidad   precio fecha_venta  
0          NaN  True                  NaN  11300.0  2018-01-01  
1         True  True                 True  69700.0  2018-02-01  
2        False  True                False  10200.0  2018-02-01  
3         True  True                  NaN  25100.0  2018-02-01  
4         True  True                 True  33400.0  2018-04-01  
...        ...   ...                  ...      ...         ...  
4838     False  True                False  14600.0  2018-08-01  
4839     False  True                 True  17500.0  2018-08-01  
4840     False  True                 True  17000.0  2018-09-01  
4841     False  True                 True  22700.0  2018-09-01  
4842     False  True                 True  15500.0  2018-09-01  

[4843 rows x 18 columns]

In [10]:
bmw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4843 entries, 0 to 4842
Data columns (total 18 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   marca                        3873 non-null   object 
 1   modelo                       4840 non-null   object 
 2   km                           4841 non-null   float64
 3   potencia                     4842 non-null   float64
 4   fecha_registro               2420 non-null   object 
 5   tipo_gasolina                4838 non-null   object 
 6   color                        4398 non-null   object 
 7   tipo_coche                   3383 non-null   object 
 8   volante_regulable            4839 non-null   object 
 9   aire_acondicionado           4357 non-null   object 
 10  camara_trasera               4841 non-null   object 
 11  asientos_traseros_plegables  1452 non-null   object 
 12  elevalunas_electrico         4841 non-null   object 
 13  bluetooth         

In [11]:
bmw.duplicated().any()

np.False_

In [12]:
bmw.isnull().any()

marca                           True
modelo                          True
km                              True
potencia                        True
fecha_registro                  True
tipo_gasolina                   True
color                           True
tipo_coche                      True
volante_regulable               True
aire_acondicionado              True
camara_trasera                  True
asientos_traseros_plegables     True
elevalunas_electrico            True
bluetooth                       True
gps                            False
alerta_lim_velocidad            True
precio                          True
fecha_venta                     True
dtype: bool

In [13]:
bmw.duplicated().sum()

np.int64(0)

### Hipotesis eliminación columnas

In [14]:
# Calcular la cantidad de valores nulos por columna
nulos_por_columna = bmw.isnull().sum()

# Calcular el porcentaje de valores nulos por columna
porcentaje_nulos = (nulos_por_columna / len(bmw)) * 100

# Calcular la cantidad de valores únicos por columna
valores_unicos_por_columna = bmw.nunique()

# Crear un nuevo DataFrame con los resultados
resultado = pd.DataFrame({
    'Cantidad de Nulos': nulos_por_columna,
    'Porcentaje de Nulos': porcentaje_nulos,
    'Valores Únicos': valores_unicos_por_columna
})

In [15]:
resultado

Cantidad de Nulos  Porcentaje de Nulos  \
marca                                      970            20.028908   
modelo                                       3             0.061945   
km                                           2             0.041297   
potencia                                     1             0.020648   
fecha_registro                            2423            50.030973   
tipo_gasolina                                5             0.103242   
color                                      445             9.188520   
tipo_coche                                1460            30.146603   
volante_regulable                            4             0.082593   
aire_acondicionado                         486            10.035102   
camara_trasera                               2             0.041297   
asientos_traseros_plegables               3391            70.018584   
elevalunas_electrico                         2             0.041297   
bluetooth                                  728            15.032005   
gps                                          0             0.000000   
alerta_lim_velocidad                       728            15.032005   
precio                                       6             0.123890   
fecha_venta                                  1             0.020648   

                             Valores Únicos  
marca                                     1  
modelo                                   76  
km                                     4784  
potencia                                 61  
fecha_registro                          180  
tipo_gasolina                             5  
color                                    10  
tipo_coche                                8  
volante_regulable                         2  
aire_acondicionado                        2  
camara_trasera                            2  
asientos_traseros_plegables               2  
elevalunas_electrico                      2  
bluetooth                                 2  
gps                                       2  
alerta_lim_velocidad                      2  
precio                                  437  
fecha_venta                              13

Ahora vamos a realizar el analisis con el porcentaje de nulos, puesto que tenemos dos outliers, como fecha_registro y asiento_traseros_plegables, en base a esto, podemos hacernos dos preguntas: eliminarlo o no eliminarlo e imputar estos nulos.

En este caso, podemos ver que en el caso de "marca" únicamente tenemos una sola marca por ello vamos a comprobar si, tenemos una marca, por ello vamos a analizar los modelos que tenemos para ver si es posible borrar la marca.

### Justificación eliminacion columnas

In [16]:
modelos_con_marca = bmw[bmw['marca'].notnull()]['modelo'].unique()

In [17]:
modelos_con_marca

array(['M4', '320', '420', '425', '335', '325', '118', '120', '220',
       '650', 'Z4', '135', '318', '430', '218', 'M3', '330', '225', '316',
       '435', '640', '640 Gran Coupé', '435 Gran Coupé', 'i8',
       '430 Gran Coupé', '635', '520', nan, '525', '530', '535',
       '320 Gran Turismo', '116', '518', '328', '523', 'X3', 'M550',
       '325 Gran Turismo', '318 Gran Turismo', '125', 'i3',
       '520 Gran Turismo', '420 Gran Coupé', '530 Gran Turismo', '114',
       'M135', '330 Gran Turismo', '418 Gran Coupé', '335 Gran Turismo',
       '730', '123', '535 Gran Turismo', '740', '750', '528',
       'ActiveHybrid 5', '735', 'M5', 'X5', 'X1', 'X4', 'X6', 'X6 M',
       'X5 M', 'X5 M50', '218 Active Tourer', '218 Gran Tourer', '216',
       '216 Gran Tourer', '216 Active Tourer', ' Active Tourer',
       '214 Gran Tourer'], dtype=object)

In [18]:
modelos_sin_marca = bmw[bmw['marca'].isnull()]['modelo'].unique()

In [19]:
modelos_sin_marca

array(['118', 'Z4', '325', '218', '120', 'M235', '425', '435', '640',
       '420', '220', 'M4', '320', '630', '318', '520', '316', '535',
       '525', '518', '530', 'M550', '330', '116', '528', '328', '125',
       '420 Gran Coupé', '114', '320 Gran Turismo', '318 Gran Turismo',
       '530 Gran Turismo', '418 Gran Coupé', '520 Gran Turismo', '135',
       '435 Gran Coupé', '640 Gran Coupé', '740', '730', 'X1', 'X3', 'X5',
       'X6', 'X4', 'X6 M', 'X5 M', 'X5 M50', '220 Active Tourer',
       '218 Active Tourer', '225 Active Tourer', '218 Gran Tourer'],
      dtype=object)

Tal y como podemos ver, los modelos que no tienen marca y los modelos que tienen marca, tienen una denominación parecida e incluso hay mucho que se repiten, en base a esta decisión vamos a eliminar la marca.

In [20]:
value_counts = bmw['color'].value_counts()

# Calcular el porcentaje de cada valor
porcentaje = (value_counts / value_counts.sum()) * 100

porcentaje

color
black     34.083674
grey      24.329241
blue      14.711232
white     11.005002
brown      6.866758
silver     6.639382
red        1.068668
beige      0.841291
green      0.318327
orange     0.136426
Name: count, dtype: float64

In [21]:
bmw['color'].isnull().sum()



np.int64(445)

In [22]:
bmw[bmw['asientos_traseros_plegables'].isnull()]

marca             modelo        km  potencia fecha_registro  \
0      NaN                118  140411.0     100.0     2012-02-01   
1      BMW                 M4   13929.0     317.0            NaN   
2      BMW                320  183297.0     120.0     2012-04-01   
3      BMW                420  128035.0     135.0            NaN   
6      BMW                325  205219.0     145.0            NaN   
...    ...                ...       ...       ...            ...   
4837   BMW    218 Gran Tourer   66770.0     110.0            NaN   
4838   BMW    218 Gran Tourer   39743.0     110.0            NaN   
4839   BMW  218 Active Tourer   49832.0     100.0     2015-06-01   
4840   BMW    218 Gran Tourer   19633.0     110.0     2015-10-01   
4842   BMW                525  195840.0     160.0            NaN   

     tipo_gasolina  color   tipo_coche volante_regulable aire_acondicionado  \
0           diesel  black          NaN              True               True   
1           petrol   grey  convertible              True               True   
2           diesel  white          NaN             False              False   
3           diesel    red  convertible              True               True   
6           diesel   grey  convertible              True               True   
...            ...    ...          ...               ...                ...   
4837        diesel   blue          van             False               True   
4838        diesel  black          NaN             False               True   
4839        diesel   grey          NaN             False               True   
4840        diesel   grey          van             False               True   
4842        diesel   grey          NaN              True                NaN   

     camara_trasera asientos_traseros_plegables elevalunas_electrico  \
0             False                         NaN                 True   
1             False                         NaN                False   
2             False                         NaN                 True   
3             False                         NaN                 True   
6             False                         NaN                 True   
...             ...                         ...                  ...   
4837          False                         NaN                False   
4838          False                         NaN                False   
4839          False                         NaN                False   
4840          False                         NaN                False   
4842          False                         NaN                 True   

     bluetooth   gps alerta_lim_velocidad   precio fecha_venta  
0          NaN  True                  NaN  11300.0  2018-01-01  
1         True  True                 True  69700.0  2018-02-01  
2        False  True                False  10200.0  2018-02-01  
3         True  True                  NaN  25100.0  2018-02-01  
6         True  True                 True  12400.0  2018-02-01  
...        ...   ...                  ...      ...         ...  
4837     False  True                False  13400.0  2018-07-01  
4838     False  True                False  14600.0  2018-08-01  
4839     False  True                 True  17500.0  2018-08-01  
4840     False  True                 True  17000.0  2018-09-01  
4842     False  True                 True  15500.0  2018-09-01  

[3391 rows x 18 columns]

In [23]:
bmw[bmw['modelo'].isnull()]

marca modelo        km  potencia fecha_registro tipo_gasolina   color  \
173    BMW    NaN  146338.0     105.0            NaN        diesel   black   
4766   BMW    NaN  115566.0     105.0     2014-01-01        diesel  silver   
4802   BMW    NaN  118640.0     110.0            NaN        diesel   white   

     tipo_coche volante_regulable aire_acondicionado camara_trasera  \
173         NaN             False               True          False   
4766        suv             False               True          False   
4802        NaN             False              False          False   

     asientos_traseros_plegables elevalunas_electrico bluetooth    gps  \
173                          NaN                False       NaN   True   
4766                         NaN                 True     False   True   
4802                         NaN                False     False  False   

     alerta_lim_velocidad   precio fecha_venta  
173                   NaN  13300.0  2018-08-01  
4766                False  14900.0  2018-09-01  
4802                False    800.0  2018-03-01

Eliminar nulos de la columna modelo, no se pude determinar cual seria

In [24]:
modelos_vacios = bmw[bmw["modelo"].isnull()].index
modelos_vacios


Index([173, 4766, 4802], dtype='int64')

In [25]:
bmw[(bmw["potencia"] == 105.0) & (bmw["tipo_gasolina"] == "diesel") & (bmw["color"] == "black")]["modelo"].unique()
# (bmw["tipo_coche"] == "suv")

array([nan, '318', '518', '118', '318 Gran Turismo', '418 Gran Coupé',
       'X1', 'X3'], dtype=object)

En base a esta respuesta, hay demasiados modelos con la misma potencia como para definir que modelo es, por ende, hemos decidido borrarlo.

In [26]:
bmw = bmw.drop(index=modelos_vacios)

### Eliminacion de columnas

Eliminaremos Marca ya que se puede asumir que son todos bmw, tambien fecha registro ya que mas de la mitad son valores nulos y no hay manera de determinarlo de otra manera, tambien eliminaria asientos_traseros_plegables por la cantidad de nulos que tiene y no se puede saber los valroes nulos.

In [27]:
bmw.drop(columns=['marca','asientos_traseros_plegables','fecha_registro'], inplace=True)

## Tratamiento duplicados

In [28]:
bmw[bmw.duplicated(keep=False)]


Empty DataFrame
Columns: [modelo, km, potencia, tipo_gasolina, color, tipo_coche, volante_regulable, aire_acondicionado, camara_trasera, elevalunas_electrico, bluetooth, gps, alerta_lim_velocidad, precio, fecha_venta]
Index: []

Por otro lado no tenemos ningun duplicado, por ello podemos avanzar.

Me he dado cuenta de que he añadido un monton de columnas con la denominación de 105, puesto que antes he pasado un .unique = 105 en vez de utilizar el doble ==, vamos a eliminarlas.

In [29]:
bmw.columns

Index(['modelo', 'km', 'potencia', 'tipo_gasolina', 'color', 'tipo_coche',
       'volante_regulable', 'aire_acondicionado', 'camara_trasera',
       'elevalunas_electrico', 'bluetooth', 'gps', 'alerta_lim_velocidad',
       'precio', 'fecha_venta'],
      dtype='object')

creamos un bucle para que itere en todas las columnas e imprima las erroneas.

In [30]:
columnas_erroneas = [col for col in bmw.columns if isinstance(col, (int, float)) or pd.isna(col)]
columnas_erroneas

[]

In [31]:
bmw = bmw.drop(columns=columnas_erroneas)

In [32]:
bmw

modelo        km  potencia tipo_gasolina   color  \
0                   118  140411.0     100.0        diesel   black   
1                    M4   13929.0     317.0        petrol    grey   
2                   320  183297.0     120.0        diesel   white   
3                   420  128035.0     135.0        diesel     red   
4                   425   97097.0     160.0        diesel  silver   
...                 ...       ...       ...           ...     ...   
4838    218 Gran Tourer   39743.0     110.0        diesel   black   
4839  218 Active Tourer   49832.0     100.0        diesel    grey   
4840    218 Gran Tourer   19633.0     110.0        diesel    grey   
4841  218 Active Tourer   27920.0     110.0        diesel   brown   
4842                525  195840.0     160.0        diesel    grey   

       tipo_coche volante_regulable aire_acondicionado camara_trasera  \
0             NaN              True               True          False   
1     convertible              True               True          False   
2             NaN             False              False          False   
3     convertible              True               True          False   
4             NaN              True               True          False   
...           ...               ...                ...            ...   
4838          NaN             False               True          False   
4839          NaN             False               True          False   
4840          van             False               True          False   
4841          van              True               True          False   
4842          NaN              True                NaN          False   

     elevalunas_electrico bluetooth   gps alerta_lim_velocidad   precio  \
0                    True       NaN  True                  NaN  11300.0   
1                   False      True  True                 True  69700.0   
2                    True     False  True                False  10200.0   
3                    True      True  True                  NaN  25100.0   
4                   False      True  True                 True  33400.0   
...                   ...       ...   ...                  ...      ...   
4838                False     False  True                False  14600.0   
4839                False     False  True                 True  17500.0   
4840                False     False  True                 True  17000.0   
4841                False     False  True                 True  22700.0   
4842                 True     False  True                 True  15500.0   

     fecha_venta  
0     2018-01-01  
1     2018-02-01  
2     2018-02-01  
3     2018-02-01  
4     2018-04-01  
...          ...  
4838  2018-08-01  
4839  2018-08-01  
4840  2018-09-01  
4841  2018-09-01  
4842  2018-09-01  

[4840 rows x 15 columns]

## Variables y valores

In [33]:
for i in bmw.columns: # Loop en las columnas del df
    print('\n',i,bmw[i].unique(),'\n') # Print nombre de la columna y valores únicos de ésta


 modelo ['118' 'M4' '320' '420' '425' '335' '325' 'Z4' '120' '220' '650' '135'
 '218' '318' '430' 'M3' '330' '225' 'M235' '316' '435' '640'
 '640 Gran Coupé' '435 Gran Coupé' 'i8' '630' '430 Gran Coupé' '635' '520'
 '525' '530' '535' '320 Gran Turismo' '518' '116' '328' '523' 'X3' 'M550'
 '528' '325 Gran Turismo' '125' '318 Gran Turismo' '420 Gran Coupé' 'i3'
 '114' '520 Gran Turismo' '530 Gran Turismo' 'M135' '418 Gran Coupé'
 '330 Gran Turismo' '335 Gran Turismo' '730' '123' '535 Gran Turismo'
 '740' '750' 'ActiveHybrid 5' '735' 'M5' 'X5' 'X1' 'X4' 'X6' 'X6 M' 'X5 M'
 'X5 M50' '218 Active Tourer' '218 Gran Tourer' '216' '220 Active Tourer'
 '216 Gran Tourer' '216 Active Tourer' ' Active Tourer'
 '225 Active Tourer' '214 Gran Tourer'] 


 km [140411.  13929. 183297. ...  19633.  27920. 195840.] 


 potencia [100. 317. 120. 135. 160. 225. 145. 105. 125. 130. 270. 110. 140. 180.
  85. 190. 236. 240. 250. 230. 309. 170. 142. 155. 210.  nan 122. 200.
  90. 150. 141. 173. 137. 280. 112. 1

Vamos a ver el número de veces que aparece cada valor de las distintas columnas del df.

In [34]:
for i in bmw.columns: # Loop en las columnas del df
    print('\n',i,bmw[i].value_counts(),'\n') 
    # Print del número de veces que aparece cada valor distinto de la columna


 modelo modelo
320                  752
520                  633
318                  568
X3                   438
116                  358
                    ... 
216                    1
220 Active Tourer      1
 Active Tourer         1
225 Active Tourer      1
214 Gran Tourer        1
Name: count, Length: 76, dtype: int64 


 km km
54433.0     2
104005.0    2
159093.0    2
180606.0    2
99768.0     2
           ..
39743.0     1
49832.0     1
19633.0     1
27920.0     1
193657.0    1
Name: count, Length: 4781, dtype: int64 


 potencia potencia
135.0    882
120.0    785
85.0     631
100.0    570
105.0    449
        ... 
412.0      1
0.0        1
423.0      1
300.0      1
261.0      1
Name: count, Length: 61, dtype: int64 


 tipo_gasolina tipo_gasolina
diesel           4628
petrol            191
hybrid_petrol       8
Diesel              5
electro             3
Name: count, dtype: int64 


 color color
black     1498
grey      1070
blue       647
white      483
brown      302
silve

In [35]:
bmw

modelo        km  potencia tipo_gasolina   color  \
0                   118  140411.0     100.0        diesel   black   
1                    M4   13929.0     317.0        petrol    grey   
2                   320  183297.0     120.0        diesel   white   
3                   420  128035.0     135.0        diesel     red   
4                   425   97097.0     160.0        diesel  silver   
...                 ...       ...       ...           ...     ...   
4838    218 Gran Tourer   39743.0     110.0        diesel   black   
4839  218 Active Tourer   49832.0     100.0        diesel    grey   
4840    218 Gran Tourer   19633.0     110.0        diesel    grey   
4841  218 Active Tourer   27920.0     110.0        diesel   brown   
4842                525  195840.0     160.0        diesel    grey   

       tipo_coche volante_regulable aire_acondicionado camara_trasera  \
0             NaN              True               True          False   
1     convertible              True               True          False   
2             NaN             False              False          False   
3     convertible              True               True          False   
4             NaN              True               True          False   
...           ...               ...                ...            ...   
4838          NaN             False               True          False   
4839          NaN             False               True          False   
4840          van             False               True          False   
4841          van              True               True          False   
4842          NaN              True                NaN          False   

     elevalunas_electrico bluetooth   gps alerta_lim_velocidad   precio  \
0                    True       NaN  True                  NaN  11300.0   
1                   False      True  True                 True  69700.0   
2                    True     False  True                False  10200.0   
3                    True      True  True                  NaN  25100.0   
4                   False      True  True                 True  33400.0   
...                   ...       ...   ...                  ...      ...   
4838                False     False  True                False  14600.0   
4839                False     False  True                 True  17500.0   
4840                False     False  True                 True  17000.0   
4841                False     False  True                 True  22700.0   
4842                 True     False  True                 True  15500.0   

     fecha_venta  
0     2018-01-01  
1     2018-02-01  
2     2018-02-01  
3     2018-02-01  
4     2018-04-01  
...          ...  
4838  2018-08-01  
4839  2018-08-01  
4840  2018-09-01  
4841  2018-09-01  
4842  2018-09-01  

[4840 rows x 15 columns]

## Tipos de variables: Target, Categoricas, Boolean, Númericas

cambio

git add "Entregable_1/Entregable1/Entregable 1.ipynb"